python/paddle/fluid/dygraph/parallel.py

```diff
@@ -424,7 +428,60 @@ def __init__(self, layers, strategy=None):
         else:
             self._strategy = _build_default_parallel_strategy()
 
+        if self._strategy.nranks > 1:
+            # convert group_size_limits MB
+            self.group_size_limits = int(group_size_limits * 1024 * 1024)
+            self.init_reducer()
+        else:
+            warnings.warn(
+                "nranks is less than 2, "
+                "maybe you need to check the current system environment."
+                " Need to use spawn or fleetrun to "
+                "start distributed programs. ")
+
+    def init_reducer(self):
+        layers_param = []
+        for sublayer in self.sublayers():
+            for _, param in sublayer.named_parameters(include_sublayers=False):
+                if not isinstance(param, core.VarBase):
+                    raise TypeError("The data type of '%s' must be Varbase" %
+                                    param.name)
+                if param.trainable:
+                    layers_param.append((sublayer, param))
+        trainable_parameters = [param for _, param in layers_param]
+
+        # NOTE(shenliang03): Here we can only use the attributes to judge whether
+        # parameter is sparse(or SelectedRows). The reason is that the sparse message
+        # can't be obtained when bp hasn't happened yet. So if layer supports sparse parameter,
+        # we should add the layer here like "nn.Embedding".
+        def check_layer_sparse(sublayer):
+            if isinstance(sublayer, nn.Embedding):
+                return sublayer._is_sparse
+            return False
+
+        is_sparse_gradient = [
+            check_layer_sparse(sublayer) for sublayer, _ in layers_param
+        ]
+        # NOTE(shenliang03): We can set environment variables to control the size of the group,
```

Comment:
Add some comments to explain what `small_group_memory` want to solve?

Reply:
done.